sirepo/pkcli/runner.py

```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
+            returncode = None
+            try:
+                returncode = await job_info.report_job.wait()
+            finally:
+                async with self.locks[run_dir]:
+                    # Clear up our in-memory status
+                    assert self.report_jobs[run_dir] is job_info
+                    del self.report_jobs[run_dir]
+                    # Write status to disk
+                    if job_info.cancel_requested:
+                        _write_status(runner_client.JobStatus.CANCELED, run_dir)
+                        await self.run_extract_job(
+                            run_dir, jhash, 'remove_last_frame', [],
+                        )
+                    elif returncode == 0:
+                        _write_status(runner_client.JobStatus.COMPLETED, run_dir)
+                    else:
                         pkdlog(
                             '{} {}: job failed, returncode = {}',
-                            run_dir, jhash, process.returncode,
+                            run_dir, jhash, returncode,
                         )
                         _write_status(runner_client.JobStatus.ERROR, run_dir)
-                    else:
-                        _write_status(runner_client.JobStatus.COMPLETED, run_dir)
-            finally:
-                # _write_status is a no-op if there's already a status, so
-                # this really means "if we get here without having written a
-                # status, assume there was some error"
-                _write_status(runner_client.JobStatus.ERROR, run_dir)
-                # Make sure that we clear out the running job info and tell
-                # everyone the job is done, no matter what happened
-                job_info = self.jobs.pop(run_dir, None)
-                if job_info is not None:
-                    job_info.finished.set()
+
+    async def run_extract_job(self, run_dir, jhash, subcmd, args):
+        pkdlog('{} {}: {} {}', run_dir, jhash, subcmd, args)
+        status = self.report_job_status(run_dir, jhash)
+        if status is runner_client.JobStatus.MISSING:
+            pkdlog('{} {}: report is missing; skipping extract job',
+                   run_dir, jhash)
+            return {}
+        # figure out which backend and any backend-specific info
+        backend_info_file = run_dir.join(_BACKEND_INFO_BASENAME)
+        if backend_info_file.exists():
+            backend_info = pkjson.load_any(backend_info_file)
+        else:
+            # Legacy run_dir
+            backend_info = pkcollections.Dict(backend='local', info={})
+
+        # run the job
+        cmd = ['sirepo', 'extract', subcmd, *args]
+        result = await _BACKENDS[backend_info.backend].run_extract_job(
+            run_dir, cmd, backend_info.info,
+        )
+
+        if result.stderr:
+            pkdlog(
+                '{} {}: got output on stderr:\n{}',
+                run_dir, jhash,
+                result.stderr.decode('utf-8', errors='ignore'),
```

Comment:
The `utf-8, errors='ignore'` is because the output should always be valid utf-8, but if something goes terribly wrong, it's better to get mangled debug output than to suffer a double-fault.

The `latin1` is because I originally forgot how to spell `utf-8, errors='ignore'`, and then after I remembered, I missed updating one of the places where I'd written it the wrong way. That's fixed now.

For writing the output, we can't make it intermingled, because `stdout` is where the structured data is. (And IIRC some of the extract jobs dump random commentary to stderr, because it's scientific code.) And see my comment above re: the chicken-and-egg problems of putting stderr into a file.

Reply:
This is the issue encapsulation should solve:

> originally forgot how to spell `utf-8, errors='ignore'`

Perhaps we need something in pkdlog. In the past, I've also avoided massive log output by limiting nested data structure recursion depth and argument size. This is a similar problem: you don't want to write control characters or other junk to the logs. If you have examples of how people encapsulate utf8 and/or these other issues, I'd appreciate it.

Try to remember to put a TODO there at some point. BTW, we use [Google's TODO mechanism](http://google.github.io/styleguide/pyguide.html#312-todo-comments): `#TODO(rjn)` or `#TODO(robnagler)` so we don't have to blame to see who put the TODO there. We do not leave a space between the `#` and `TODO` to allow simple grepping. Not a big deal, just fix next SOW.